Fix 'set_storage_pool' and 'set_snapshot_pool' methods in Volumes

aalexmonteiro opened the pull request:
### Description
Fix 'set_storage_pool' method for all APIS and 'set_snapshot_pool' method for API 500 in Volumes.

### Issues Resolved
#279 

### Check List
- [ ] New functionality includes testing.
  - [x] All tests pass (`$ rake test`).
- [ ] New functionality has been documented in the README if applicable.
  - [ ] New functionality has been thoroughly documented in the examples (please include helpful comments).
- [x] Changes are documented in the CHANGELOG.

tmiotto commented on lib/oneview-sdk/resource/api200/volume.rb:
```diff
@@ -68,11 +68,7 @@ def set_storage_system(storage_system)
-        if self['provisioningParameters']
-          self['provisioningParameters']['storagePoolUri'] = storage_pool['uri']
-        else
-          self['storagePoolUri'] = storage_pool['uri']
```
You've ommited this part in the new code. There is a reason why?

aalexmonteiro replied:
Because Ricardo commented that it is not possible to update a storage pool, I checked this in the documentation. Do you think the previous form is better?

tmiotto replied:
You are saying you cannot update by changing the Storage Pool associated with the volume, right?

aalexmonteiro replied:
Yes! This property is read only.

tmiotto commented on lib/oneview-sdk/resource/api500/c7000/volume.rb:
```diff
@@ -62,16 +62,19 @@ def delete(flag = :all)
           assure_uri(storage_pool)
-          @data['properties'] ||= {}
-          @data['properties']['storagePool'] = storage_pool['uri']
+          @data['properties']['storagePool'] = storage_pool['uri'] if @data['properties']
```
As far as it fixes the PUT it may break the POST.

tmiotto replied:
I'm concerned if it won't cause any breakage in already existing code

aalexmonteiro replied:
I understand. Maybe another way would be inside the update method, delete the properties attribute. What do you think?

ricardogpsf replied:
Why not keeping the code how have been working before?
Because this helper method is just to be used before of the create (POST) action. For update, the user cannot set the storage_pool.
Maybe documenting that be sufficient.

ricardogpsf commented on lib/oneview-sdk/resource/api500/c7000/volume.rb:
```diff
@@ -62,16 +62,19 @@ def delete(flag = :all)
           assure_uri(storage_pool)
-          @data['properties'] ||= {}
-          @data['properties']['snapshotPool'] = storage_pool['uri']
+          if @data['properties']
```
Maybe can be better check if this resource will be used to create a new resource or update the current resource, we can check if the 'uri' attribute is set.
And we should document this behavior.

aalexmonteiro replied:
But if the user is trying to create something that already exists or in the case of the 'get_volume_template_uri' method I can assign a storage_pool without the properties attribute? I think it better to do the update treatment as it is done in the create.

ricardogpsf replied:
> But if the user is trying to create something that already exists 

I did not understand the reason for this.

> the case of the 'get_volume_template_uri' method I can assign a storage_pool without the properties attribute?

but the `get_volume_template_uri` method seems getting the value from correct place:
```ruby
 def get_volume_template_uri(template_data)
   storage_pool_uri = self['storagePoolUri'] || self['properties']['storagePool']
   ...
```

aalexmonteiro replied:
Ok! Maybe that can solve. ```@data['properties'] ||= {} unless @data['uri']```

Because it does not make much sense for me to check the existence of a volume in this method. If uri was passed, I assume user knows that the resource already exists.

ricardogpsf commented on spec/integration/shared_examples/volume/api200/update_snapshot_pool.rb:
```diff
@@ -0,0 +1,32 @@
+  describe '#update snapshot pool' do
+    it 'updating the snapshot pool' do
+      old_snapshot_pool = resource_class_of('StoragePool').find_by(current_client, uri: item['snapshotPoolUri']).first
+      new_snapshot_pool = resource_class_of('StoragePool').get_all(current_client).first
```
It can occur of same storage_pool be returned by those two calls?

ricardogpsf commented on spec/integration/shared_examples/volume/api500/create.rb:
```diff
@@ -231,18 +231,30 @@
       item = described_class.new(current_client, name: VOLUME_NAME)
       item.set_storage_pool(storage_pool)
-      expect(item['properties']['storagePool']).to eq(storage_pool['uri'])
+      expect(item['storagePoolUri']).to eq(storage_pool['uri'])
```
But the result of `set_storage_pool ` should have set the `['properties']['storagePool']`, because it is the test of create action, right?! For create action, the URI of storage pool should be inside `item['properties']['storagePool']`.

ricardogpsf replied:
no problem. That things happen. 😄

ricardogpsf commented on spec/integration/shared_examples/volume/api500/create.rb:
```diff
@@ -231,18 +231,30 @@
       item = described_class.new(current_client, name: VOLUME_NAME)
       item.set_snapshot_pool(storage_pool)
-      expect(item['properties']['snapshotPool']).to eq(storage_pool['uri'])
+      expect(item['snapshotPoolUri']).to eq(storage_pool['uri'])
```
same case of my last comment.

ricardogpsf commented on spec/integration/shared_examples/volume/create.rb:
```diff
@@ -112,10 +112,16 @@
       item = described_class.new(current_client, name: VOLUME_NAME)
       item.set_storage_pool(storage_pool)
-      expect(item['provisioningParameters']['storagePoolUri']).to eq(storage_pool['uri'])
+      expect(item['storagePoolUri']).to eq(storage_pool['uri'])
```
same case of my last comment.

ricardogpsf commented on spec/unit/resource/api500/c7000/volume_spec.rb:
```diff
@@ -87,19 +87,31 @@
-    it 'sets the snapshotPool attribute' do
-      item = described_class.new(@client_500)
+    it 'sets the snapshotPool attribute in properties' do
+      item = described_class.new(@client_500, properties: {})
```
if the user are not pass the properties as an empty `Hash` and set the storate_pool, will the create action create a Volume with that storage pool associated? If not, is this behavior correct?
For me, no problem use this new approach, but maybe important documenting that over the constructor of the Volume class.

ricardogpsf commented on spec/unit/resource/api500/c7000/volume_spec.rb:
```diff
@@ -87,19 +87,31 @@
     end
+
+    it 'sets the snapshotPoolUri attribute' do
+      item = described_class.new(@client_500, deviceSpecificAttributes: {})
```
same case of my last comment.

aalexmonteiro commented:
@ricardogpsf , I removed one comment, which was a replica of another comment I made, but then I saw that it did not make sense. So I removed mine and to make it cleaner, I also removed yours.
